diff --git a/base/plugins.cpp b/base/plugins.cpp
index 5d0be11..a259f83 100644
--- a/base/plugins.cpp
+++ b/base/plugins.cpp
@@ -103,6 +103,9 @@ public:
 		#if PLUGIN_ENABLED_STATIC(CRUISE)
 		LINK_PLUGIN(CRUISE)
 		#endif
+		#if PLUGIN_ENABLED_STATIC(DARKSEED2)
+		LINK_PLUGIN(DARKSEED2)
+		#endif
 		#if PLUGIN_ENABLED_STATIC(DRACI)
 		LINK_PLUGIN(DRACI)
 		#endif
diff --git a/configure b/configure
index db01465..3d472ed 100755
--- a/configure
+++ b/configure
@@ -83,6 +83,7 @@ add_engine agos "AGOS" yes "agos2"
 add_engine agos2 "AGOS 2 games" yes
 add_engine cine "Cinematique evo 1" yes
 add_engine cruise "Cinematique evo 2" yes
+add_engine darkseed2 "Dark Seed II" no
 add_engine draci "Dragon History" yes
 add_engine drascula "Drascula: The Vampire Strikes Back" yes
 add_engine gob "Gobli*ns" yes
@@ -2048,7 +2049,7 @@ define_in_config_if_yes "$_build_hq_scalers" 'USE_HQ_SCALERS'
 #
 if test "$_indeo3" = auto ; then
 	# Autodetect. Build if either the gob engine or plugins are enabled
-	if test `get_engine_build gob` = yes || test "$_dynamic_modules" = yes ; then
+	if test `get_engine_build gob` = yes || test `get_engine_build darkseed2` = yes || test "$_dynamic_modules" = yes ; then
 		_indeo3=yes
 	else
 		_indeo3=no
diff --git a/engines/engines.mk b/engines/engines.mk
index 2c13782..a4bd523 100644
--- a/engines/engines.mk
+++ b/engines/engines.mk
@@ -36,6 +36,11 @@ DEFINES += -DENABLE_CRUISE=$(ENABLE_CRUISE)
 MODULES += engines/cruise
 endif
 
+ifdef ENABLE_DARKSEED2
+DEFINES += -DENABLE_DARKSEED2=$(ENABLE_DARKSEED2)
+MODULES += engines/darkseed2
+endif
+
 ifdef ENABLE_DRACI
 DEFINES += -DENABLE_DRACI=$(ENABLE_DRACI)
 MODULES += engines/draci
